pandas/core/missing.py

```diff
@@ -225,41 +225,25 @@ def interpolate_1d(
     # default limit is unlimited GH #16282
     limit = algos._validate_limit(nobs=None, limit=limit)
 
-    # These are sets of index pointers to invalid values... i.e. {0, 1, etc...
-    all_nans = set(np.flatnonzero(invalid))
-    start_nans = set(range(find_valid_index(yvalues, "first")))
-    end_nans = set(range(1 + find_valid_index(yvalues, "last"), len(valid)))
-    mid_nans = all_nans - start_nans - end_nans
-
-    # Like the sets above, preserve_nans contains indices of invalid values,
-    # but in this case, it is the final set of indices that need to be
-    # preserved as NaN after the interpolation.
-
-    # For example if limit_direction='forward' then preserve_nans will
-    # contain indices of NaNs at the beginning of the series, and NaNs that
-    # are more than'limit' away from the prior non-NaN.
-
-    # set preserve_nans based on direction using _interp_limit
-    preserve_nans: Union[List, Set]
+    first = find_valid_index(yvalues, "first")
+    last = find_valid_index(yvalues, "last")
     if limit_direction == "forward":
-        preserve_nans = start_nans | set(_interp_limit(invalid, limit, 0))
+        nans_to_interpolate = _interp_limit(invalid, limit, 0, first, last)
     elif limit_direction == "backward":
-        preserve_nans = end_nans | set(_interp_limit(invalid, 0, limit))
+        nans_to_interpolate = _interp_limit(invalid, 0, limit, first, last)
     else:
         # both directions... just use _interp_limit
-        preserve_nans = set(_interp_limit(invalid, limit, limit))
+        nans_to_interpolate = _interp_limit(invalid, limit, limit, first, last)
```

Comment:
we could move the above logic into _interp_limit and fastpath if limit_area.

although the algorithm could probably be extended to 2d, in which case the limit_area logic below would also need to move.

Reply:
> although the algorithm could probably be extended to 2d, in which case the limit_area logic below would also need to move.

i don't think a 2d version will be needed as interpolate_2d will also be applied along axis when max_gap is added. 

> we could move the above logic into _interp_limit and fastpath if limit_area.

probably best as a follow-on to keep the diff here smaller